enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,401 @@
+	w := &workloadIdentityCredential{file: file}
+	cred, err := azidentity.NewClientAssertionCredential(tenantID, clientID, w.getAssertion, &azidentity.ClientAssertionCredentialOptions{ClientOptions: options.ClientOptions})
+	if err != nil {
+		return nil, err
+	}
+	w.cred = cred
+	return w, nil
+}
+
+func (w *workloadIdentityCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return w.cred.GetToken(ctx, opts)
+}
+
+func (w *workloadIdentityCredential) getAssertion(context.Context) (string, error) {
+	if now := time.Now(); w.lastRead.Add(5 * time.Minute).Before(now) {
+		content, err := os.ReadFile(w.file)
+		if err != nil {
+			return "", err
+		}
+		w.assertion = string(content)
+		w.lastRead = now
+	}
+	return w.assertion, nil
+}
+
+func getAuthorizerForResource(config Config) (autorest.Authorizer, error) {
+  ...
+
+	var (
+		cred azcore.TokenCredential
+		err  error
+	)
+
+  // ClientSecret is absent AND FederatedTokenFile has been set, create a workloadIdentityCredential
+	if config.ClientSecret == "" && config.FederatedTokenFile != "" {
+		options := workloadIdentityCredentialOptions{
+			ClientOptions: azcore.ClientOptions{
+				Cloud: cloudConfig,
+			},
+		}
+		cred, err = newWorkloadIdentityCredential(config.TenantID, config.ClientID, config.FederatedTokenFile, &options)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		options := azidentity.ClientSecretCredentialOptions{
+			ClientOptions: azcore.ClientOptions{
+				Cloud: cloudConfig,
+			},
+		}
+		cred, err = azidentity.NewClientSecretCredential(config.TenantID, config.ClientID, config.ClientSecret, &options)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	scope := config.Environment.TokenAudience
+	if !strings.HasSuffix(scope, "/.default") {
+		scope += "/.default"
+	}
+	// Use an adapter so azidentity in the Azure SDK can be used as
+	// Authorizer when calling the Azure Management Packages, which we
+	// currently use. Once the Azure SDK clients (found in /sdk) move to
+	// stable, we can update our clients and they will be able to use the
+	// creds directly without the authorizer. The schedule is here:
+	// https://azure.github.io/azure-sdk/releases/latest/index.html#go
+	authorizer := azidext.NewTokenCredentialAdapter(cred, []string{scope})
+	return authorizer, nil
+}
+```
+
+#### Mutating admission webhook
+
+CCO will deploy and lifecycle the [Azure Workload Identity mutating admission webhook](https://azure.github.io/azure-workload-identity/docs/installation/mutating-admission-webhook.html) on Azure clusters such that customers can annotate workload `ServiceAccounts` with Managed Identity details necessary for creating clients. When the mutating admission webhook finds these annotations on a
+`ServiceAccount` referenced by a pod being created, environment variables are set for the pod for the `AZURE_CLIENT_ID`, `AZURE_TENANT_ID` and `AZURE_FEDERATED_TOKEN_FILE`.
```

Comment:
What will this webhook's failure mode be? Discuss.

Reply:
Likely `failurePolicy: Ignore`, see: https://issues.redhat.com/browse/CCO-259 for discussion as it related to the AWS pod identity webhook.

Reply:
> Because -- just making sure -- if the webhook doesn't work, there's no recourse, right? In particular, there's no falling back to the old auth method, because those creds will simply not exist anywhere?

If the webhook isn't available or doesn't work then the credentials just won't be there as environment variables but the distinction here is that OpenShift's operators will not utilize the mutating webhook, at least not at first, and will instead continue to use config secrets.